docs/design/2022-06-22-foreign-key.md

```diff
@@ -0,0 +1,720 @@
+  `id` int NOT NULL,
+  `a` int DEFAULT NULL,
+  PRIMARY KEY (`id`),
+  KEY `fk` (`a`),
+  CONSTRAINT `t_ibfk_1` FOREIGN KEY (`a`) REFERENCES `t` (`id`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
+1 row in set
+```
+
+As you can see, the index `fk` is auto create for foreign key usage.
+
+#### Validate
+
+Create a table with foreign key, check following condition when build DDL job and DDL owner received DDL job(aka Double-Check):
+
+- whether the user have `REFERENCES` privilege to the foreign key references table.
+- Corresponding columns in the foreign key and the referenced key must have similar data types. The size and sign of fixed precision types such as INTEGER and DECIMAL must be the same. The length of string types need not be the same. For nonbinary (character) string columns, the character set and collation must be the same.
+- Supports foreign key references between one column and another within a table. (A column cannot have a foreign key reference to itself.)
+- Require indexes on foreign keys and referenced keys so that foreign key checks can be fast and not require a table scan. The size and sign of fixed precision types such as INTEGER and DECIMAL must be the same. The length of string types need not be the same. For nonbinary (character) string columns, the character set and collation must be the same.
+- Index prefixes on foreign key columns are not supported. Consequently, BLOB and TEXT columns cannot be included in a foreign key because indexes on those columns must always include a prefix length.
+- Does not currently support foreign keys for tables with user-defined partitioning. This includes both reference and child tables.
+- A foreign key constraint cannot reference a virtual generated column, but stored generated column is ok.
+
+#### Handle In DDL Owner
+
+When DDL Owner handle create table job, DDL owner need to create a new table, and update the reference tables information.
+
+At the same point in time, there may be two versions of the schema in the TiDB cluster, so we can't create new table and 
+update all reference tables in one schema version, since this may break foreign key constrain, such as delete reference table
+without foreign key constrain check in child table.
+
+```sql
+-- In TiDB-1 and Schema Version is 1
+insert into t_has_foreign_key values (1, 1);
+
+-- In TiDB-0 and  Schema Version is 0
+delete from t_reference where id = 1; --Since doesn't know foreign key information in old version, so doesn't do foreign key constrain check.
+```
+
+So, when create a table with foreign key, we need multi-schema version change:
+
+1. None -> Write Only: Create table with state is `write-only`, and update all reference tables info.
+2. Write Only -> Done: Update the new created table state to `public`.
+
+In step-1, we need update some table info in one schema-version. Technically, we can implement is since we already support `ActionCreateTables` DDL job.
```

Comment:
```suggestion
In step-1, we need update some table info in one schema version. Technically, we can implement it since we already support `ActionCreateTables` DDL job.
```